### DISCUSSION TITLE
How to customize sorting of candidates sifted by company-capf?

### DISCUSSION THREAD

**Discussion by kings2u** [company-mode/company-mode#1223] [Q&A]
Despite studying the code of `company-capf`, I cannot determine how to write a custom completion-at-point-function that contains the necessary metadata for the `company-capf` function to take advantage of the `company-capf--sorted` variable offered in `company-capf.el`. There is no keyword like `:company-docsig` or `:company-kind` as there is with other parts of `company-capf`, though customizing sorting is clearly possible from what I can see in the `company-capf--candidates` function. Any advice on how to construct my CAPF would be great!

**Comment by dgutov** [comment]
There is no `:company-*` for it because sorting is an original feature of a completion table, as defined in the core.

See the docstring of `completion-metadata` and `display-sort-function` in particular. You'll probably want to ask how to assign this metadata field to a completion table -- hopefully the description at the top of `lisp/minibuffer.el` can answer that. Do you know how to find it?

And yes, it seems unnecessarily complex, but we have what we have.

**Reply by kings2u** [reply]
Thanks for the pointers! And yes, the critical issue I can’t answer is how to assign the metadata to a completion table in the first place for use in a CAPF. I’m unable to locate `medatata.el` with `find-library` or `locate`. Did you mean `minibuffer.el`? 

The best resource I’ve found on the topic of adding metadata to a table is https://with-emacs.com/posts/tutorials/customize-completion-at-point/ but its example deals with the context of `completing-read`, not completion-at-point-functions. Another similar resource is [here](https://stackoverflow.com/questions/36799299/emacs-sort-files-in-completions-buffer-when-opening-a-file). FYI, I’m trying to combine my own list of possible completion candidates with all those in `obarray` and sort them from shortest length to longest.

Do you have any suggestions?

**Reply by dgutov** [reply]
> I’m unable to locate medatata.el with find-library or locate. Did you mean minibuffer.el?

Yes. Sorry! I'll edit the answer above.

**Reply by dgutov** [reply]
> Do you have any suggestions?

Here's a concrete example: `project--file-completion-table` in `lisp/progmodes/project.el`. It's a function that takes a flat list of strings and returns a closure which works as a completion table, and which reports `((category . project-file))` as its metadata.

**Reply by dgutov** [reply]
Ah, and the example at with-emacs.com also creates such completion table.

The difference with c-a-p-f is that a `completion-at-point-functions` element is another function that needs to return a list like:

```
(START END TABLE ...EXTRA-PROPS)
```

Where TABLE is a completion table (often a function) which can be defined like described above.

**Comment by kings2u** [comment]
Thank you for clarifying. With your pointers, and after examining the code for `eglot.el`, I was able to narrow down the basic template to sort candidates like I wanted. :)
```
;; -*- lexical-binding: t -*-
(setq test-candidates '("aaaaa" "aaaaaaaaaa" "aaa" "aaaaaaaa"))
(defun test-capf ()
  (let* ((start (save-excursion (skip-syntax-backward "w_")
                               (point)))
         (end (point))
         (sort-completions (lambda (completions)
                            (sort completions
                                  (lambda (x y)
                                       (or (< (length x) (length y))
                                           (and (= (length x) (length y))
                                                (string< x y)))))))
         (metadata `(metadata . ((display-sort-function . ,sort-completions)))))
    (when (and start end)
      (list start end
            test-candidates 
            :company-kind (lambda (cand) 'variable)))))
(setq-local completion-at-point-functions '(test-capf t))
```

**Reply by dgutov** [reply]
The posted code snippet doesn't seem to be using the `metadata` variable anywhere.

And `test-candidates` is a list of string literals instead of a function, which is necessary for it to be able to report metadata.

So I'm not sure how this solves your problem.

**Reply by kings2u** [reply]
I’ve updated my answer to make clear that I believe lexical binding is required. If you paste the code into a new buffer and evaluate it and type "aaa", you will see that it sorts the `test-candidates` in order of string length even though they are out of order in the original list. 

I honestly don’t fully understand how it works myself 😅 But as @joaotavora made clear [in a similar issue](https://github.com/joaotavora/eglot/issues/190#issuecomment-449637774), the built-in emacs completion framework is quite complex under the hood. I think the `metadata` variable comes into play when `test-capf` is passed to emacs' internal completion functions like `all-completions`?

**Reply by dgutov** [reply]
> I think the metadata variable comes into play when test-capf is passed to emacs' internal completion functions like all-completions?

There is no global variable called `metadata`. And if you try byte-compiling your file, the compilation has a linter step which will report that the variable is unused.

> If you paste the code into a new buffer and evaluate it and type "aaa", you will see that it sorts the test-candidates in order of string length even though they are out of order in the original list.

That's the default behavior: unless a backend reports that it's `sorted`, the completions are sorted alphabetically. And `(sort '("aaaaa" "aaaaaaaaaa" "aaa" "aaaaaaaa") #'string<)` evaluates to the order which takes lengths into account as well.

You can delete the `sort-completions` and `metadata` declarations, and the behavior will remain the same.

**Reply by kings2u** [reply]
Oh my, I’m so sorry. I’m clearly in over my head here 😅 Thanks for pointing all that out, I'll keep trying to learn what’s going on.

**Reply by kings2u** [reply]
Here is the buffer I'm evalling with comments where appropriate: 
```
;; -*- lexical-binding: t -*-
(setq test-candidates '("aaaa" "aaaaaaaaaaaa" "aaaaa" "aaaaaaa"))

;; I've confirmed the following works to sort the candidates in order of longest to shortest strings:
(defun sort-completions (completions)
  (sort completions
        (lambda (x y)
          (or (> (length x) (length y))
              (and (= (length x) (length y))
                   (string> x y))))))

;; I’ve modified the `project--file-completion-table` function so it will (hopefully) create a table with metadata of a `display-sort-function` that will use my `sort-completions` function:
(defun my-project--file-completion-table (all-files)
  (lambda (string pred action)
    (cond
     ((eq action 'metadata)
      `(metadata . ((display-sort-function . sort-completions)))) ; NOTE  eglot-completion-at-point has a comma in front of its sort-completions function...
     (t
      (complete-with-action action all-files string pred)))))
;; following seems to confirm that `sort-completions` was successfully stored as `display-sort-function` metadata in the table:
;; (completion-metadata-get (completion-metadata
;;                           "aaa"
;;                           (my-project--file-completion-table test-candidates) nil)
;;                          'display-sort-function)
;;    --> sort-completions

;; Finally I created a sample c-a-p-f that tries and fails to complete against `test-candidates`:
(defun test-capf ()
  (let* ((start (save-excursion (skip-syntax-backward "w_")
                               (point)))
         (end (point)))
    (when (and start end)
      (list start
            end
            (my-project--file-completion-table test-candidates)))))

(setq-local completion-at-point-functions '(test-capf t))
```

If I `eval-buffer`, then typing "aaa" produces no error but fails to sort the list from longest string to shortest in the company tooltip. (If I eval each form one by one instead of one `eval-buffer`, "aaa" produces an error that `all-files` is a void variable, but I’m assuming that’s because with lexical binding you have to eval the whole buffer at once?)

However, if I change this line in `company-capf--candidates`
```
(sortfun (cdr (assq 'display-sort-function meta)))
```
to this 
```
(sortfun (completion-metadata-get (completion-metadata
                                    input table pred) 
                          'display-sort-function))
```
my custom sorting works as expected. 

I have seen this `completion-metadata-get` method in several packages like Embark, Vertico, Marginalia, Ivy, and `minibuffer.el`. I’m not sure if you think my personal functions should be changed somehow or if `company-capf--candidates` should be updated. If the former, what changes would you suggest to accomplish the custom sorting?

**Reply by dgutov** [reply]
Hmm, I just copied your code, it didn't do what's expected right away. I fiddled with print-debugging and then evaluated this:

```el
(setq company-transformers nil)
```

And then it worked as expected: completions are sorted from longest to shortest.

Perhaps your personal config also adds something to `company-transformers`, to customize sorting?

If you're still having problems, you can add some print-debugging to `company-capf--candidates` like this:

```diff
diff --git a/company-capf.el b/company-capf.el
index 813b2aa..210ab7c 100644
--- a/company-capf.el
+++ b/company-capf.el
@@ -186,11 +186,15 @@ so we can't just use the preceding variable instead.")
              (sortfun (cdr (assq 'display-sort-function meta)))
              (last (last candidates))
              (base-size (and (numberp (cdr last)) (cdr last))))
+        (message "meta: %s" meta)
+        (message "sortfun: %s" sortfun)
         (when base-size
           (setcdr last nil))
         (setq company-capf--sorted (functionp sortfun))
+        (message "candidates before: %s" candidates)
         (when sortfun
           (setq candidates (funcall sortfun candidates)))
+        (message "candidates after: %s" candidates)
         (if (not (zerop (or base-size 0)))
             (let ((before (substring input 0 base-size)))
               (mapcar (lambda (candidate)
```

Apply the change, evaluate the buffer and see how the values transform.

> However, if I change this line in company-capf--candidates <...>

Aren't those the same? See `completion-metadata-get`'s definition: it just does `(cdr (assq prop metadata))`. I don't see any difference in behavior after applying this.

Note that `company-capf` caches completions, so to see the changes work you need to restart completion (dismiss the popup with `C-g` and start again). But you probably tried that already.

**Reply by dgutov** [reply]
BTW, to jump to a function's definition, move point to its name and press `M-.`.

**Reply by kings2u** [reply]
> Perhaps your personal config also adds something

You were right about this. I had this flex-styling snippet in my config, and removing it made the sorting function take effect:
```
(define-advice company-capf 
    (:around (orig-fun &rest args) set-completion-styles)
  (let ((completion-styles '(flex)))
    (apply orig-fun args)))
```
However, I’m not sure that tells the whole story since the `completion-metadata-get` method I described does seem to allow my `advice` to coexist with my custom `sort-completions` function:
> Aren't those the same? See completion-metadata-get's definition: it just does (cdr (assq prop metadata)). I don't see any difference in behavior after applying this.

Doing the debugging you suggested does suggest a subtle difference: the `(sortfun (cdr (assq 'display-sort-function meta)))` line produces a `sortfun` with the byte-coded function value of my `sort-completions` function, whereas the alternative `completion-metadata-get` method produces a `sortfun` value of the plain symbol `sort-completions`. 

Does this information raise a solution as to how I might tailor my `sort-completions` function while keeping my flex-styling `advice` to `company-capf`? Perhaps I should just use the rewritten `company-capf--candidates` in my personal config? It would be nice to have the best of both worlds: flex completion styles _and_ custom sorting.

**Reply by dgutov** [reply]
> Does this information raise a solution as to how I might tailor my sort-completions function while keeping my flex-styling advice to company-capf?

Not yet. I don't see the difference in those implementations.

> (sortfun (cdr (assq 'display-sort-function meta))) line produces a sortfun with the byte-coded function value of my sort-completions function, whereas the alternative completion-metadata-get method produces a sortfun value of the plain symbol sort-completions

Does this mean that to reproduce this I need to byte-compile the file with your custom code first?

**Reply by kings2u** [reply]
> Does this mean that to reproduce this I need to byte-compile the file with your custom code first?

I meant that `(sortfun (cdr (assq 'display-sort-function meta)))` sets `sortfun` with the value of a byte-coded function object as described [here](https://www.gnu.org/software/emacs/manual/html_node/elisp/Byte_002dCode-Type.html). So when the critical line `(setq candidates (funcall sortfun candidates))` is executed, `sortfun` has the value:
```
#[257 \300\203
\0\300!\202\0\211\301 \203\0d\302 V\203\0\303\304"\202\0\211\207 [sort-completions window-minibuffer-p minibuffer-prompt-end sort #[514 \300\301\302#\300\301\302#\206\0\301\206\0\301V\207 [get-text-property 0 completion-score] 7 
(fn C1 C2)]] 5 
(fn COMPLETIONS)] 
```
whereas with the `completion-metadata-get` method, `sortfun` just has the value `sort-completions`.

I have debugged the current version of `company-capf--candidates` with the following changes
```diff
(setq company-capf--sorted (functionp sortfun))
- (when sortfun
-            (setq candidates (funcall sortfun candidates)))
+ (when sortfun
+          (let ((inhibit-redisplay nil))
+            (debug)
+            (setq candidates (funcall sortfun candidates))))
```
and observed how my `sort-completions` function initially works to reorder the candidates, but then another unnamed function (whose source code is not available to me) jumps in to re-sort them back into the default order. 

Anyways, thanks for your patient help as always :) I can mark this as solved if you like since it seems you’re content with the current behavior.

**Reply by kings2u** [reply]
Also interesting is that changing my `advice` to `company-capf` to use `orderless-flex` instead of the built-in emacs `flex` resolves the problem without needing to change any part of `company-capf--candidates`:
```
(define-advice company-capf
    (:around (orig-fun &rest args) set-completion-styles)
  (let ((completion-styles '(orderless))
        (orderless-matching-styles '(orderless-flex)))
    (apply orig-fun args)))
```

**Reply by dgutov** [reply]
> I meant that (sortfun (cdr (assq 'display-sort-function meta))) sets sortfun with the value of a byte-coded function object as described here.

Hmm.

I guess it's a sorting function added by the `flex` style. It's unconditionally added in `completion--flex-adjust-metadata`. It's probably "the right thing" for us to honor this.

The `completion-metadata-get` might actually get the wrong result, leading to no flex sorting in the general scenario as well. I wonder how, though. It's as if `input` and `(buffer-substring (nth 1 res) (nth 2 res))` are different, but I can't reproduce that with print-debugging.
